conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/CaffeineCachingTypeFactory.java

```diff
@@ -0,0 +1,145 @@
+            TypeParser parser,
+            @Nullable TypeModifier[] modifiers,
+            ClassLoader classLoader) {
+        super(typeCache, parser, modifiers, classLoader);
+    }
+
+    private static CaffeineLookupCache asCaffeine(LookupCache<Object, JavaType> typeCache) {
+        if (typeCache instanceof CaffeineLookupCache) {
+            return (CaffeineLookupCache) typeCache;
+        } else if (typeCache == null || typeCache.size() == 0) {
+            // Use a caffeine cache instead of the provided empty cache
+            return new CaffeineLookupCache();
+        } else if (typeCache instanceof LRUMap) {
+            CaffeineLookupCache cache = new CaffeineLookupCache();
+            LRUMap<Object, JavaType> defaultJacksonCacheType = (LRUMap<Object, JavaType>) typeCache;
+            defaultJacksonCacheType.contents(cache::put);
+            return cache;
+        } else {
+            log.error(
+                    "Unknown LookupCache, using a new caffeine cache instead",
+                    SafeArg.of("cacheType", typeCache.getClass()));
+            return new CaffeineLookupCache();
+        }
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withModifier(TypeModifier mod) {
+        // Updating modifiers clears the cache
+        return new CaffeineCachingTypeFactory(
+                new CaffeineLookupCache(), _parser, computeModifiers(_modifiers, mod), _classLoader);
+    }
+
+    @Nullable
+    private static TypeModifier[] computeModifiers(TypeModifier[] existing, TypeModifier newModifier) {
+        // Semantics are based on the jackson-databind `withModifier` implementation.
+        if (newModifier == null) {
+            return null;
+        }
+        if (existing == null || existing.length == 0) {
+            return new TypeModifier[] {newModifier};
+        }
+        return ArrayBuilders.insertInListNoDup(existing, newModifier);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withClassLoader(ClassLoader classLoader) {
+        return new CaffeineCachingTypeFactory(asCaffeine(_typeCache), _parser, _modifiers, classLoader);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withCache(LRUMap<Object, JavaType> cache) {
+        return withCache((LookupCache<Object, JavaType>) cache);
+    }
+
+    @Override
+    public CaffeineCachingTypeFactory withCache(LookupCache<Object, JavaType> cache) {
+        return new CaffeineCachingTypeFactory(asCaffeine(cache), _parser, _modifiers, _classLoader);
+    }
+
+    private static final class CaffeineLookupCache implements LookupCache<Object, JavaType> {
+        private final Cache<Object, JavaType> cache;
+
+        CaffeineLookupCache() {
+            this.cache = Caffeine.newBuilder()
+                    // max-size 1000 up from 200 default as of 2.14.2
+                    .maximumSize(1000)
+                    // initial-size 128 up from 16 default as of 2.14.2
+                    .initialCapacity(128)
```

Comment:
I'd really like to add a max-age here to allow eventual class unloading, but it's not clear what a good value would be.

Reply:
Should we use an expireAfterRead of relatively short duration?

Reply:
That's what I was had in mind, sorry, not sure why I said 'max-age' 🤦

The caffeine cache implementation appears equivalent to the cache used in jackson, where contention occurs within ConcurrentHashMaps, so my current theory is that we have a hot rotating set of types beyond the 200 element size of the cache. Since endpoints are called frequently, we end up churning the same items in and out. Each time an item is expired, several threads attempt to recompute and re-add it to the cache in parallel.